File: src/roundedbuttons/RoundedButtons.java
package roundedbuttons;

import lib.io.FastReader;

import java.util.ArrayList;
import java.util.List;

import static roundedbuttons.Pair.*;

public class RoundedButtons {

    public static void main(String[] args) {
        String in = """
                3
                1 8 14 13 3 5 8 6 15 8 12 11 14 21 2.5 20
                2 2 13 14 4 5 4 4 4 16 13 14 3 3 3 9
                85.7 114.7 3.2 6.0 1.2 3 86.3 114.8 88.1 118.2 85.9 120.7
                """;
        FastReader io = new FastReader(in);
        int testCases = Integer.parseInt(io.nextLine());

        double x, y, w, h, r;
        for (int i = 0; i < testCases; i++) {
            x = io.nextDouble();
            y = io.nextDouble();
            w = io.nextDouble();
            h = io.nextDouble();
            r = io.nextDouble();

            List<Pair> clicks = new ArrayList<>();
            int numPairs = io.nextInt();
            for (int k = 0; k < numPairs; k++) {
                clicks.add(new Pair(io.nextDouble(), io.nextDouble()));
            }

            List<String> result = calcInOut(x, y, w, h, r, clicks);
            for (String str : result) {
                System.out.println(str);
            }
            System.out.println();
        }
    }

    static List<String> calcInOut(double x, double y, double w, double h, double r, List<Pair> clicks) {
        List<String> result = new ArrayList<>();
        double inWidth  = w - (2 * r);
        double inHeight = h - (2 * r);

        // calculate all 4 center points
        Pair center2 = new Pair(x + r, y + r);
        Pair center3 = new Pair(x + r, y + r + inHeight);
        Pair center1 = new Pair(x + r + inWidth, y + r);
        Pair center4 = new Pair(x + r + inWidth, y + r + inHeight);

        double xBegin = x + r;
        double xEnd = xBegin + inWidth;
        double yEnd = y + h;

        for (Pair click : clicks) {
            if (click.x >= xBegin && click.x <= xEnd) {     // vertical rectangle
                if (click.y >= y && click.y <= yEnd) {
                    result.add("inside");
                    continue;
                }
            } else if (click.x >= x && click.x <= x + w) {  // horizontal rectangle
                if (click.y >= y + r && click.y <= y + r + inHeight) {
                    result.add("inside");
                    continue;
                }
            }

            if (click.x < x + r) {
                if (click.y < y + r) {      // quadrant 2
                    populateArray(center2, click, r, result);
                } else {                    // quadrant 3
                    populateArray(center3, click, r, result);
                }
            } else {
                if (click.y < y + r) {      // quadrant 1
                    populateArray(center1, click, r, result);
                } else {                    // quadrant 4
                    populateArray(center4, click, r, result);
                }
            }
        }
        return result;
    }
}

class Pair {
    double x;
    double y;

    public Pair(double x, double y) {
        this.x = x;
        this.y = y;
    }

    static double calcDistance(Pair center, Pair click) {
        double xDistance = (center.x - click.x);
        double yDistance = (center.y - click.y);
        double x = Math.pow(xDistance, 2);
        double y = Math.pow(yDistance, 2);

        return Math.pow(x + y, 0.5);
    }

    static boolean isInside(double distance, double radius) {
        return distance <= radius;
    }

    static void populateArray(Pair center, Pair click, double radius, List<String> result) {
        double distance = calcDistance(center, click);
        if (isInside(distance, radius)) {
            result.add("inside");
        } else {
            result.add("outside");
        }
    }
}

File: src/roundedbuttons/roundedbuttons.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"strconv"
	"strings"
)

func main() {
	input := "roundedbuttons/in"
	file, err := os.Open(input)

	if err != nil {
		fmt.Printf("File not found: '%v'", input)
		return
	}

	io := bufio.NewScanner(file)
	io.Scan()
	cases, _ := strconv.Atoi(io.Text())

	for i := 0; i < cases; i++ {
		io.Scan()
		words := strings.Fields(io.Text())

		rect := words[0:5]
		x, _ := strconv.ParseFloat(rect[0], 64)
		y, _ := strconv.ParseFloat(rect[1], 64)
		w, _ := strconv.ParseFloat(rect[2], 64)
		h, _ := strconv.ParseFloat(rect[3], 64)
		r, _ := strconv.ParseFloat(rect[4], 64)
		var pt Point
		rect1 := Rect{x, y, w, h, r, pt}

		clicks := words[6:]

		var click []Point
		for i := 0; i < len(clicks); i += 2 {
			x, _ := strconv.ParseFloat(clicks[i], 64)
			y, _ := strconv.ParseFloat(clicks[i+1], 64)
			click = append(click, Point{x, y})
		}
		calcInOut(&rect1, &click)
		fmt.Println()
	}
}

func calcInOut(rect *Rect, clicks *[]Point) {
	inWidth  	:= rect.w - (2 * rect.r)
	inHeight	:= rect.h - (2 * rect.r)
	xBegin 		:= rect.x + rect.r
	xEnd 		:= rect.x + rect.r + inWidth
	yBegin  	:= rect.y + rect.r
	yEnd 		:= rect.y + rect.h

	center1 := Point{xBegin, yBegin}
	center2 := Point{xBegin, yBegin + inHeight}
	center3 := Point{xBegin + inWidth, yBegin}
	center4 := Point{xBegin + inWidth, yBegin + inHeight}

	for _, pt := range *clicks {
		if pt.x >= xBegin && pt.x <= xEnd {						// horizontal rect
			if pt.y >= rect.y && pt.y <= yEnd {
				fmt.Println("inside")
				continue
			}
		} else if pt.x >= rect.x && pt.x <= rect.x + rect.w {	// vertical rect
			if pt.y >= yBegin && pt.y <= yBegin + inHeight {
				fmt.Println("inside")
				continue
			}
		}

		mid := rect.midPoint()			// calc the center of the rectangle (x, y)
		if pt.x <= mid.x {
			if pt.y <= mid.y {			// bottom left
				dist := center1.calcDistance(&pt)
				printResult(rect, dist)
			} else {					// top left
				dist := center2.calcDistance(&pt)
				printResult(rect, dist)
			}
		} else if pt.y <= mid.y {		// bottom right
			dist := center3.calcDistance(&pt)
			printResult(rect, dist)
		} else {						// top right
			dist := center4.calcDistance(&pt)
			printResult(rect, dist)
		}
	}
}

func (center *Point) calcDistance(click *Point) float64 {
	xDist := center.x - click.x
	yDist := center.y - click.y
	x := math.Pow(xDist, 2)
	y := math.Pow(yDist, 2)

	return math.Pow(x+y, 0.5)
}

func printResult(rect *Rect, dist float64) {
	if dist <= rect.r {
		fmt.Println("inside")
	} else {
		fmt.Println("outside")
	}
}

type Point struct {
	x, y float64
}

type Rect struct {
	x, y, w, h, r float64
	midpoint      Point
}

func (r *Rect) midPoint() Point {
	x := r.w/2 + r.x
	y := r.h/2 + r.y
	return Point{x, y}
}
